mono/metadata/unity-liveness.c

```diff
@@ -1,100 +1,150 @@
 #include <config.h>
 #include <glib.h>
-#include <mono/metadata/object.h>
-#include <mono/metadata/object-internals.h>
-#include <mono/metadata/metadata.h>
-#include <mono/metadata/tabledefs.h>
 #include <mono/metadata/class-internals.h>
 #include <mono/metadata/domain-internals.h>
+#include <mono/metadata/metadata.h>
+#include <mono/metadata/object-internals.h>
+#include <mono/metadata/object.h>
+#include <mono/metadata/tabledefs.h>
 #include <mono/utils/mono-error.h>
 
 typedef struct _LivenessState LivenessState;
 
-typedef struct _GPtrArray custom_growable_array;
-#define array_at_index(array,index) (array)->pdata[(index)]
+#define k_block_size (8 * 1024)
+#define k_array_elements_per_block ((k_block_size - 3 * sizeof (void*)) / sizeof (gpointer))
 
-#if defined(HAVE_SGEN_GC)
-void sgen_stop_world (int generation);
-void sgen_restart_world (int generation);
-#elif defined(HAVE_BOEHM_GC)
-#ifdef HAVE_BDWGC_GC
-extern void GC_stop_world_external();
-extern void GC_start_world_external();
-#else
-void GC_stop_world_external()
-{
-	g_assert_not_reached ();
-}
-void GC_start_world_external()
-{
-	g_assert_not_reached ();
-}
-#endif
-#else
-#error need to implement liveness GC API
-#endif
+typedef struct _custom_array_block custom_array_block;
+
+typedef struct _custom_array_block {
+	gpointer *next_item;
+	custom_array_block *prev_block;
+	custom_array_block *next_block;
+	gpointer p_data[k_array_elements_per_block];
+} custom_array_block;
 
-custom_growable_array* array_create_and_initialize (guint capacity)
+typedef struct _custom_block_array_iterator custom_block_array_iterator;
+
+typedef struct _custom_growable_block_array {
+	custom_array_block *first_block;
+	custom_array_block *current_block;
+	custom_block_array_iterator *iterator;
+} custom_growable_block_array;
+
+typedef struct _custom_block_array_iterator {
+	custom_growable_block_array *array;
+	custom_array_block *current_block;
+	gpointer *current_position;
+} custom_block_array_iterator;
+
+
+typedef void(*register_object_callback) (gpointer *arr, int size, void *callback_userdata);
+typedef void(*WorldStateChanged) ();
+typedef void *(*ReallocateArray) (void *ptr, int size, void *callback_userdata);
+
+struct _LivenessState {
+	custom_growable_block_array *all_objects;
+
+	MonoClass *filter;
+
+	custom_growable_block_array *process_array;
+	guint initial_alloc_count;
+
+	void *callback_userdata;
+
+	register_object_callback filter_callback;
+	ReallocateArray reallocateArray;
+	guint traverse_depth; // track recursion. Prevent stack overflow by limiting recursion
+};
+
+custom_growable_block_array * block_array_create(LivenessState *state)
 {
-	custom_growable_array* array = g_ptr_array_sized_new(capacity);
-	array->len = 0;
+	custom_growable_block_array *array = g_new0(custom_growable_block_array, 1);
+	array->current_block = state->reallocateArray(NULL, k_block_size, state->callback_userdata);
```

Comment:
So the constraint here is that the world is stopped, this reallocateArray callback will never acquire a lock of any kind that may be held by another suspended thread?

Reply:
yes. The unity side ensures no locks are taken, that could be held by the stopped threads